### ISSUE TITLE
JUNK tech: ai

### ISSUE THREAD

**Issue by Whyisthisnotavalable** [landgreen/n-gon#74]
(function() {
    const inputSize = 8; 
    const hiddenSize = 16;
    const outputSize = 4;

    const weightsInputHidden = new Array(inputSize).fill(0).map(() => new Array(hiddenSize).fill(0));
    const biasesHidden = new Array(hiddenSize).fill(0);
    const weightsHiddenOutput = new Array(hiddenSize).fill(0).map(() => new Array(outputSize).fill(0));
    const biasesOutput = new Array(outputSize).fill(0);
    // Initialize weights with small random values between -0.5 and 0.5
    for (let i = 0; i < inputSize; i++) {
        for (let j = 0; j < hiddenSize; j++) {
            weightsInputHidden[i][j] = Math.random() - 0.5;
        }
    }

    for (let i = 0; i < hiddenSize; i++) {
        for (let j = 0; j < outputSize; j++) {
            weightsHiddenOutput[i][j] = Math.random() - 0.5;
        }
    }
    for (let i = 0; i < hiddenSize; i++) {
        biasesHidden[i] = Math.random() - 0.5;
    }

    for (let i = 0; i < outputSize; i++) {
        biasesOutput[i] = Math.random() - 0.5;
    }
    function sigmoid(x) {
        return 1 / (1 + Math.exp(-x));
    }
    function neuralNetwork(input) {
        const hiddenLayer = new Array(hiddenSize).fill(0);
        for (let i = 0; i < hiddenSize; i++) {
            for (let j = 0; j < inputSize; j++) {
                hiddenLayer[i] += input[j] * weightsInputHidden[j][i];
            }
            hiddenLayer[i] = sigmoid(hiddenLayer[i] + biasesHidden[i]);
        }
        const output = new Array(outputSize).fill(0);
        for (let i = 0; i < outputSize; i++) {
            for (let j = 0; j < hiddenSize; j++) {
                output[i] += hiddenLayer[j] * weightsHiddenOutput[j][i];
            }
            output[i] = sigmoid(output[i] + biasesOutput[i]);
        }
        return output;
    }
    function updateInputArray(player, exitX, exitY, mobPositions, mapPositions, bodyPositions) {
        const parsedMobPositions = JSON.parse(mobPositions);
        const mobX = parsedMobPositions.map(position => position.x);
        const mobY = parsedMobPositions.map(position => position.y);
    
        const parsedMapPositions = JSON.parse(mapPositions);
        const mapX = parsedMapPositions.map(position => position.x);
        const mapY = parsedMapPositions.map(position => position.y);
    
        const parsedBodyPositions = JSON.parse(bodyPositions);
        const bodyX = parsedBodyPositions.map(position => position.x);
        const bodyY = parsedBodyPositions.map(position => position.y);
    
        const input = [
            player.position.x,
            player.position.y,
            exitX - player.position.x,
            exitY - player.position.y,
            ...mobX,
            ...mobY,
            ...mapX,
            ...mapY,
            ...bodyX,
            ...bodyY
        ];
        return input;
    }    
    function updateNeuralNetwork(player, exitX, exitY, mobPositions, mapLength, bodyLength) {
        const inputt = updateInputArray(player, exitX, exitY, mobPositions, mapLength, bodyLength);
        const output = neuralNetwork(inputt);
        const [up, down, left, right, field, fire] = output;
        console.log(output)
        input.up = up > 0.5;
        input.down = down > 0.5;
        input.left = left > 0.5;
        input.right = right > 0.5;
        input.field = field > 0.5;
        input.fire = fire > 0.5;
    }
    function loop() {
        const mobPos = [];
        for(let i = 0; i < mob.length; i++) {
            mobPos.push({x: mob[i].position.x, y: mob[i].position.y})
        }
        const mapPos = [];
        for(let i = 0; i < map.length; i++) {
            mapPos.push({x: map[i].position.x, y: map[i].position.y})
        }
        const bodyPos = [];
        for(let i = 0; i < body.length; i++) {
            bodyPos.push({x: body[i].position.x, y: body[i].position.y})
        }
        updateNeuralNetwork(player, level.exit.x, level.exit.y, JSON.stringify(mobPos), JSON.stringify(mapPos), JSON.stringify(bodyPos));
        requestAnimationFrame(loop);
    }
    loop();
})();

**Comment by Whyisthisnotavalable** [comment]
Yields very funny results

**Comment by landgreen** [comment]
what does it do?

**Comment by Whyisthisnotavalable** [comment]
Ai sort of thing. Try it by pasting it in the console.

**Comment by landgreen** [comment]
I did, it made it hard to move.

**Comment by Whyisthisnotavalable** [comment]
try it multiple times, it should be different each time

**Comment by landgreen** [comment]
what does it do?

**Comment by Whyisthisnotavalable** [comment]
It is supposedly a neural network that takes the player position mob positions, body, etc and gives a result

**Comment by Whyisthisnotavalable** [comment]
Technically, you could train it to beat the game, but it just gives it a random bias each time

**Comment by Whyisthisnotavalable** [comment]
[recording.webm](https://github.com/landgreen/n-gon/assets/77077277/605874c0-96ca-4b4f-9ed6-79b6fe8d05e2)

**Comment by Whyisthisnotavalable** [comment]
This is where it controls the angle

**Comment by landgreen** [comment]
well... I don't have a response.

**Comment by Whyisthisnotavalable** [comment]
(function() {
const inputSize = 13;
const hiddenSize = 16;
const outputSize = 7;

const weightsInputHidden = new Array(inputSize).fill(0).map(() => new Array(hiddenSize).fill(0));
const biasesHidden = new Array(hiddenSize).fill(0);
const weightsHiddenOutput = new Array(hiddenSize).fill(0).map(() => new Array(outputSize).fill(0));
const biasesOutput = new Array(outputSize).fill(0);
for (let i = 0; i < inputSize; i++) {
    for (let j = 0; j < hiddenSize; j++) {
        weightsInputHidden[i][j] = Math.random() - 0.5;
    }
}
for (let i = 0; i < hiddenSize; i++) {
    for (let j = 0; j < outputSize; j++) {
        weightsHiddenOutput[i][j] = Math.random() - 0.5;
    }
}
for (let i = 0; i < hiddenSize; i++) {
    biasesHidden[i] = Math.random() - 0.5;
}

for (let i = 0; i < outputSize; i++) {
    biasesOutput[i] = Math.random() - 0.5;
}
function sigmoid(x) {
    return 1 / (1 + Math.exp(-x));
}
function neuralNetwork(input) {
    const hiddenLayer = new Array(hiddenSize).fill(0);
    for (let i = 0; i < hiddenSize; i++) {
        for (let j = 0; j < inputSize; j++) {
            hiddenLayer[i] += input[j] * weightsInputHidden[j][i];
        }
        hiddenLayer[i] = sigmoid(hiddenLayer[i] + biasesHidden[i]);
    }
    const output = new Array(outputSize).fill(0);
    for (let i = 0; i < outputSize; i++) {
        for (let j = 0; j < hiddenSize; j++) {
            output[i] += hiddenLayer[j] * weightsHiddenOutput[j][i];
        }
        output[i] = sigmoid(output[i] + biasesOutput[i]);
    }
    return { output, hiddenLayer };
}
const explorationFactor = 1000;

function ucbScore(actionValue, actionCount, totalSteps) {
    const explorationBonus = explorationFactor * Math.sqrt(Math.log(totalSteps + 1) / (actionCount + 1));
    return actionValue + explorationBonus;
}
function calculateDesiredOutput(player, exitX, exitY) {
    const desiredOutput = [0, 0, 0, 0];
    
    if (player.position.y > exitY) {
        desiredOutput[0] = 1;
    } else if (player.position.y < exitY) {
        desiredOutput[1] = 1;
    }
    
    if (player.position.x > exitX) {
        desiredOutput[2] = 1; 
    } else if (player.position.x < exitX) {
        desiredOutput[3] = 1; 
    }
    
    return desiredOutput;
}
function calculateLoss(desiredOutput, actualOutput) {
    let loss = 0;
    for (let i = 0; i < desiredOutput.length; i++) {
        loss += Math.pow(desiredOutput[i] - actualOutput[i], 2);
    }
    return loss;
}
function updateInputArray(player, exitX, exitY, mobPositions, mapPositions, bodyPositions) {
    const parsedMobPositions = JSON.parse(mobPositions);
    const mobX = parsedMobPositions.map(position => position.x);
    const mobY = parsedMobPositions.map(position => position.y);

    const parsedMapPositions = JSON.parse(mapPositions);
    const mapX = parsedMapPositions.map(position => position.x);
    const mapY = parsedMapPositions.map(position => position.y);

    const parsedBodyPositions = JSON.parse(bodyPositions);
    const bodyX = parsedBodyPositions.map(position => position.x);
    const bodyY = parsedBodyPositions.map(position => position.y);

    const input = [
        m.health,
        /* m.angle, */
        player.velocity.x,
        player.velocity.y,
        player.position.x,
        player.position.y,
        exitX - player.position.x,
        exitY - player.position.y,
        ...mobX,
        ...mobY,
        ...mapX,
        ...mapY,
        ...bodyX,
        ...bodyY
    ];
    console.log(input);
    return input;
}    
const learningRate = 0.01;
function updateWeightsAndBiases(desiredOutput, actualOutput, hiddenLayer) {
    const outputErrors = [];
    for (let i = 0; i < desiredOutput.length; i++) {
        outputErrors[i] = (desiredOutput[i] - actualOutput[i]) * actualOutput[i] * (1 - actualOutput[i]);
    }
    for (let i = 0; i < hiddenSize; i++) {
        for (let j = 0; j < outputSize; j++) {
            weightsHiddenOutput[i][j] += learningRate * hiddenLayer[i] * outputErrors[j];
        }
        biasesOutput[i] += learningRate * outputErrors[i];
    }
}

function updateNeuralNetwork(player, exitX, exitY, mobPositions, mapPositions, bodyPositions) {
    const inputt = updateInputArray(player, exitX, exitY, mobPositions, mapPositions, bodyPositions);
    const { output, hiddenLayer } = neuralNetwork(inputt);

    const desiredOutput = calculateDesiredOutput(player, exitX, exitY);
    const loss = calculateLoss(desiredOutput, output);

    updateWeightsAndBiases(desiredOutput, output, hiddenLayer);
    const totalSteps = m.cycle;
    const ucbValues = output.map((actionValue, index) => ucbScore(actionValue, totalSteps, index));

    const [up, down, left, right, field, fire, angle] = output;
    input.up = up > 0.5;
    input.down = down > 0.5;
    input.left = left > 0.5;
    input.right = right > 0.5;
    input.field = field > 0.5;
    input.fire = fire > 0.5;
    /* m.angle = angle * Math.PI * 2; */
}

function loop() {
    /* m.look = () => {
        const scale = 0.8;
        m.transSmoothX = canvas.width2 - m.pos.x - (simulation.mouse.x - canvas.width2) * scale;
        m.transSmoothY = canvas.height2 - m.pos.y - (simulation.mouse.y - canvas.height2) * scale;

        m.transX += (m.transSmoothX - m.transX) * 0.07;
        m.transY += (m.transSmoothY - m.transY) * 0.07;
    }; */
    const mobPos = [];
    for(let i = 0; i < mob.length; i++) {
        mobPos.push({x: mob[i].position.x, y: mob[i].position.y})
    }
    const mapPos = [];
    for(let i = 0; i < map.length; i++) {
        mapPos.push({x: map[i].position.x, y: map[i].position.y})
    }
    const bodyPos = [];
    for(let i = 0; i < body.length; i++) {
        bodyPos.push({x: body[i].position.x, y: body[i].position.y})
    }
    updateNeuralNetwork(player, level.exit.x, level.exit.y, JSON.stringify(mobPos), JSON.stringify(mapPos), JSON.stringify(bodyPos));
    requestAnimationFrame(loop);
}
loop();
})();

**Comment by Whyisthisnotavalable** [comment]
This could be an actual tech, as it seems to work a little bit better

**Comment by landgreen** [comment]
...

**Comment by Whyisthisnotavalable** [comment]
Did you try it?

**Comment by Cdev-99** [comment]
I'ma try making a new version of the ai
